docs/architecture/adr-102-rpc-companion.md

```diff
@@ -0,0 +1,344 @@
+
+##### Data types
+
+This solution will implement a data schema in the database using Postgresql built-in data types. By using a relational database, there's a possibility
+to better normalize the data structures, this might provide savings in storage but might also add to the complexity of returning a particular
+dataset because the data joins that will be required. Also, it would be important ensure the referential integrity is not violated since
+this can cause issues to the clients consuming the data.
+
+In order to accurately ensure the database is storing the full node data types properly, the database might implement custom data types (`domains` in Postgresql).
+
+For example, Postgresql doesn't have an unsigned `uint64` datatype, therefore in order to support this in the database, you can use a `domain`, which is a base types with additional constrains. For example, this is the definition of a `uint64` domain:
+
+```sql
+-- DOMAIN: comet.uint64
+
+CREATE DOMAIN comet.uint64
+    AS numeric;
+
+ALTER DOMAIN comet.uint64
+    ADD CONSTRAINT value_max CHECK (VALUE <= '18446744073709551615'::numeric);
+
+ALTER DOMAIN comet.uint64
+    ADD CONSTRAINT value_positive CHECK (VALUE >= 0::numeric);
+```
+Additional domains for `uint32` and `uint8` will also be implemented.
+
+##### Schema migration
+
+Another point to consider is when the data structures change across CometBFT releases. There are a couple of ways to provide a solution to this
+problem.
+
+One way is to support a mechanism to migrate the old data to the new data structures. This would require additional logic for the migration process
+that would need to be run before ingesting data from an upgraded full node that contains the new data structures. But this approach might also
+cause some issues, for example, if the new data structure has a new field that there's no corresponding value in the "old" data structure, probably
+the value would need to be set to `[null]` but this can have unintended consequences.
+
+Another potential solution for this scenario is to find a way in the database that can support "versioning" of data structures. For example, let's
+assume there's a `Block` structure, let's call it `v1`. If in the future there's a need to modify this structure that is not compatible with the
+previous data structure, then the database would support a `v2` schema for `ResultBlock` and an `index` table could determine the criteria on which
+data structure should be used for inserting or querying data.
+
+There's also a possibility that the data structure could be stored in parallel with the normalized data but in one blob, e.g. in a `jsonb` field
+that contains all the information needed to return the client, this could be an optimized way to serve the data but could add to the storage
+requirements. This could improve the performance for reading data from the database. An alternative approach to this problem that can offer
+similar results would be to use a "caching" layer.
```

Comment:
Thanks Andy for explaining all these options. My only worry with separate versioned data is support for queries (somehow you'd need to query more than one table (or not? ) etc.) Maybe we could do the first thing proposed (copy the old one into the new structure), but add a version column and perhaps do something with the fields that differ on query response? (like if its `v1` and the field is `null` put a default value, or remove it ? 

Also, note that returning `nil` in the fields that did not exist before might be needed, because if people want to query data both in the old format and new format, they can't receive entries in different formats, its either one or the other. We could support both but as a user, I'd find that harder to manage. I'd rather have everything in one format with a version field indicating that I can then do some extra parsing after I get the results.

Reply:
We might need to thinks this through and see what's the best way to approach this. I'm wondering how this might have been handled in CometBFT in the past, like the data structures changing and how that was handled from the RPC perspective.

Reply:
since we now have proto versioning, I like the idea of using the same versioning as the proto files since the db schema should match the proto structure. This might be the easiest way to handle data structure changes.